ui/src/Roles/RoleClient.ts

```diff
@@ -38,40 +41,58 @@ class RoleClient implements TraitClient {
         return Axios.get(url, config);
     }
 
-    async add(role: RoleAddRequest, spaceUuid: string): Promise<AxiosResponse> {
-        const url = this.getBaseRolesUrl(spaceUuid);
+    async add(addRequest: TraitAddRequest, space: Space): Promise<AxiosResponse<Trait>> {
+        const url = this.getBaseRolesUrl(space.uuid!!);
         let config = {
             headers: {
                 'Content-Type': 'application/json',
                 'Authorization': `Bearer ${getToken()}`,
             },
         };
 
-        return Axios.post(url, role, config);
+        return Axios.post(url, addRequest, config).then((result) => {
+            MatomoEvents.pushEvent(space.name, "addRole", addRequest.name);
+            return result;
+        }).catch((err) => {
+            MatomoEvents.pushEvent(space.name, "addRoleError", addRequest.name, err.code);
+            return Promise.reject(err);
+        });
     }
 
-    async edit(role: RoleEditRequest, spaceUuid: string): Promise<AxiosResponse> {
-        const url = this.getBaseRolesUrl(spaceUuid);
+    async edit(editRequest: TraitEditRequest, space: Space): Promise<AxiosResponse<Trait>> {
+        const url = this.getBaseRolesUrl(space.uuid!!);
         let config = {
             headers: {
                 'Content-Type': 'application/json',
                 'Authorization': `Bearer ${getToken()}`,
             },
         };
 
-        return Axios.put(url, role, config);
+        return Axios.put(url, editRequest, config).then((result) => {
+            MatomoEvents.pushEvent(space.name, "editRole", editRequest.updatedName!!);
+            return result;
```

Comment:
I'm getting a 400 bad request when I try to edit a role..

Reply:
You need to restart your back end some object got modified